wiki/Client Security.md

```diff
@@ -0,0 +1,95 @@
+
+- Argon2 has a [browser-based implementation](https://github.com/antelle/argon2-browser) based on WebAssembly. It offers satisfactory performance under reasonably low memory settings, and can complete within a reasonable amount of time for 10<sup>6</sup> repeatitions. Therefore, it would not slow down the setup process
+- Based on our knowledge, no one has built an ASIC miner for computing Argon2 hashes at this time. 
+- Argon2 is not yet even popular enough to be included in [hashcat](https://hashcat.net/hashcat/), the well-known GPU/FPGA based password recovery tool by brute-forcing hash computations via GPU.
+
+<a name="f1">[1]</a> See an outdated [Litecoin Mining Hardware Comparison](https://litecoin.info/index.php/Mining_hardware_comparison)
+
+<a name="f2">[2]</a> See [AntMiner L3+](https://shop.bitmain.com/promote/antminer_l3_litecoin_asic_scrypt_miner/specification) which achieves 504MH/s
+
+#### III. Scrambled Memory Layout
+
+We can further improve the security of **GenLeaf** by placing the result in the memory where it would be difficult for the attacker to locate, but would be easy for the client with the correct OTP to retrieve. A naive construct following this intuition is to store each leaf at the position of the OTP they correspond to and a pointer to their neighbor. Unfortunately this naive approach leaks the OTP code of each leaf and renders OTP useless. A mild improvement from the naive approach is to swap each leaf with its neighbor's position (after first repositioning all leaves to their OTP values) and remove the pointer to the neighbor (since neighbor is part of the Merkle Proof, not the leaf node itself). This construct no longer leaks the OTP for each leaf, but an attacker would still be unable to identify the location for each pair of leaves by pre-comptuing the parent hash of all pairs of leaves, which can be done in M<sup>2</sup> operations and would only require 32 bytes * M amount of memory. Since each operation is a SHA256 hash, based on the analysis in Section I, an attacker with a high-end GPU only needs ~10 minutes to identify exact positions of all leaves before they are scrambled. Moreover, we cannot use the techniques in Section II by replacing the hash function to a complex one, since the hash function must be supported on blockchain as well and must be economical and fast enough to compute on blockchain.
+
+These two naive approaches are unsatisfactory because they are missing a few critical elements:
+
+1. They did not use randomness.
+2. Their position mapping function for each leaf is invertible.
+3. They did not expand the search space for brute-force attacker (confined within M<sup>2</sup>)
+
+To address these issues, we introduce the following technique:
+
+1. For each leaf, we use the corresponding OTP as the seed for a random number generator, to generate 4 consecutive numbers between [1, 4M]. Note this mapping is deterministic (hence repeatable) but is not invertible.
```

Comment:
From @ivan-homoliak-sutd :
> why consecutive? I'd say random.

Reply:
Should be rephrased to `to consecutively generate 4 random numbers between [1, 4M]`. The consecutive requirement is for determinism, so that a user can restore the entire wallet construction and scrambled memory layout using only the same OTP seed.